rllib/agents/trainer.py

```diff
@@ -742,11 +749,35 @@ def env_creator_from_classpath(env_context):
         self.local_replay_buffer = (
             self._create_local_replay_buffer_if_necessary(self.config))
 
-        # Make the call to self._init. Sub-classes should override this
-        # method to implement custom initialization logic.
-        self._init(self.config, self.env_creator)
-
-        # Evaluation setup.
+        # Deprecated way of implementing Trainer sub-classes (or "templates"
+        # via the soon-to-be deprecated `build_trainer` utility function).
+        # Instead, sub-classes should override the Trainable's `setup()`
+        # method and call super().setup() from within that override at some
+        # point.
+        self.workers = None
+        self.train_exec_impl = None
+        # Old design: Override `_init`
+        try:
+            self._init(self.config, self.env_creator)
+        # New design: Override `setup` (as indented by Trainable)
+        # and do or don't call super().setup() from within your override.
+        # By default, `setup` should create both worker sets: "rollout workers"
+        # for collecting samples for training and - if applicable - "evaluation
+        # workers".
+        except NotImplementedError:
```

Comment:
I am curious, instead of this try except, we should be able to simply put all the logics in this except block into a default self._init() function?
that seems to have the same effect as this.

Reply:
Yeah, great point. Let me fix this.

Reply:
Not sure. We are trying to deprecate the `_init()` method (in favor of `Trainable.setup()` as indented by Tune Trainable API). Moving functionality now into `_init` just to avoid the try/except would counter that effort and again make it harder to read the code in `setup()` (use would have to jump into `_init` to find the WorkerSet generating code).

Reply:
yeah, I figured it doesn't hurt to ask :) thanks for the explanation.